### PR TITLE
Workflow support

### LINKED ISSUE #682: Support for ensemble of models served in torchserve
## Is your feature request related to a problem? Please describe.
- Currently to implement ensemble of models, the user needs to create a pipeline and manage data flow on the client side.
- This introduces an overhead of multiple API requests, slowing down the overall output.

## Describe the solution
- Ensembling could be achieved at the server side by providing a simple configuration which represents the pipeline for data flow.
- This will reduce the number of requests being sent by the end user and should result result in faster overall output

### PR DESCRIPTION
**Pull request by dhaniram-kshirsagar**
## Description
Please refer issue [comment](https://github.com/pytorch/serve/issues/682#issuecomment-718063289) for details

Fixes #682 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the tests [UT/IT] that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced. 
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- UT/IT execution results

- Logs

## Checklist:

- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] New and existing unit tests pass locally with these changes?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by dhanainme** [comment]
These ~150 files have been split to the following branches for more managable 



- issue_682_workflow_archiver_frontend                     293ec798 Frontend Ensemble Code
- issue_682_workflow_archiver_frontend_test                a35f1699 Frontend Java Workflow Archive Tests
- issue_682_workflow_archiver_py                           ba7a9d12 Workflow Archiver Tool
- issue_682_workflow_backend                               11b62dab Workflow backend changes
- issue_682_workflow_scripts                               8ead3c30 Workflow - Updates to scripts
- issue_682_workflow_server_docs_examples                  675cef1b Workflow updates to Docs & Examples
- issue_682_workflow_server_ensemble                       a6e1362c Gradle changes
- issue_682_workflow_server_http                           1c3b5d78 Frontend Java code changes for HTTP / GRPC
- issue_682_workflow_server_test                           0fdbcf06 Workflow test updates
- issue_682_workflow_server_wlm_utils                      ea214155 Gradle / Server updates
- issue_682_workflow_server_workflowcode                   d5ad80f1 Java Frontend Server Workflow Code



**issue_682_workflow_server_docs_examples**

M       README.md
M       docs/README.md
M       docs/configuration.md
M       docs/rest_api.md
M       docs/server.md
A       docs/workflow_inference_api.md
A       docs/workflow_management_api.md
A       docs/workflows.md
A       examples/Workflows/README.md
A       examples/Workflows/densenet_image_classifier_pipeline/README.md
R100    frontend/modelarchive/src/test/resources/models/noop_no_archive/noop/model.pt   examples/Workflows/densenet_image_classifier_pipeline/densenet_model/README.md
A       examples/Workflows/densenet_image_classifier_pipeline/densenet_model/densenet_handler.py
A       examples/Workflows/densenet_image_classifier_pipeline/densenet_model/model.py
A       examples/Workflows/densenet_image_classifier_pipeline/densenet_workflow.yaml
A       examples/Workflows/densenet_image_classifier_pipeline/densenet_workflow_handler.py
A       examples/Workflows/densenet_image_classifier_pipeline/index_to_name.json
A       examples/Workflows/densenet_image_classifier_pipeline/kitten.jpg



**issue_682_workflow_archiver_frontend**

R100    frontend/modelarchive/build.gradle      frontend/archive/build.gradle
R100    frontend/modelarchive/gradlew   frontend/archive/gradlew
R084    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/DownloadModelException.java       frontend/archive/src/main/java/org/pytorch/serve/archive/DownloadArchiveException.java
R096    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/InvalidModelException.java        frontend/archive/src/main/java/org/pytorch/serve/archive/model/InvalidModelException.java
R098    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/Manifest.java     frontend/archive/src/main/java/org/pytorch/serve/archive/model/Manifest.java
R058    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/ModelArchive.java frontend/archive/src/main/java/org/pytorch/serve/archive/model/ModelArchive.java
R096    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/ModelException.java       frontend/archive/src/main/java/org/pytorch/serve/archive/model/ModelException.java
R096    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/ModelNotFoundException.java       frontend/archive/src/main/java/org/pytorch/serve/archive/model/ModelNotFoundException.java
R096    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/ModelVersionNotFoundException.java        frontend/archive/src/main/java/org/pytorch/serve/archive/model/ModelVersionNotFoundException.java
A       frontend/archive/src/main/java/org/pytorch/serve/archive/utils/ArchiveUtils.java
R095    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/HexUtils.java     frontend/archive/src/main/java/org/pytorch/serve/archive/utils/HexUtils.java
A       frontend/archive/src/main/java/org/pytorch/serve/archive/utils/InvalidArchiveURLException.java
R067    frontend/modelarchive/src/main/java/org/pytorch/serve/archive/ZipUtils.java     frontend/archive/src/main/java/org/pytorch/serve/archive/utils/ZipUtils.java
A       frontend/archive/src/main/java/org/pytorch/serve/archive/workflow/InvalidWorkflowException.java
A       frontend/archive/src/main/java/org/pytorch/serve/archive/workflow/Manifest.java
A       frontend/archive/src/main/java/org/pytorch/serve/archive/workflow/WorkflowArchive.java
A       frontend/archive/src/main/java/org/pytorch/serve/archive/workflow/WorkflowException.java
A       frontend/archive/src/main/java/org/pytorch/serve/archive/workflow/WorkflowNotFoundException.java



**issue_682_workflow_archiver_frontend_test**

A       frontend/archive/src/test/java/org/pytorch/serve/archive/CoverageTest.java
R082    frontend/modelarchive/src/test/java/org/pytorch/serve/archive/ModelArchiveTest.java     frontend/archive/src/test/java/org/pytorch/serve/archive/model/ModelArchiveTest.java
A       frontend/archive/src/test/java/org/pytorch/serve/archive/workflow/WorkFlowArchiveTest.java
R100    frontend/modelarchive/src/test/java/org/pytorch/serve/test/TestHelper.java      frontend/archive/src/test/java/org/pytorch/serve/test/TestHelper.java
R100    frontend/modelarchive/src/test/resources/models/custom_invalid_python_dep.mar   frontend/archive/src/test/resources/models/custom_invalid_python_dep.mar
R100    frontend/modelarchive/src/test/resources/models/custom_python_dep.mar   frontend/archive/src/test/resources/models/custom_python_dep.mar
R100    frontend/modelarchive/src/test/resources/models/echo.mar        frontend/archive/src/test/resources/models/echo.mar
R100    frontend/modelarchive/src/test/resources/models/error_batch.mar frontend/archive/src/test/resources/models/error_batch.mar
R100    frontend/modelarchive/src/test/resources/models/init-error.mar  frontend/archive/src/test/resources/models/init-error.mar
R100    frontend/modelarchive/src/test/resources/models/loading-memory-error.mar        frontend/archive/src/test/resources/models/loading-memory-error.mar
R100    frontend/modelarchive/src/test/resources/models/mnist.mar       frontend/archive/src/test/resources/models/mnist.mar
R100    frontend/modelarchive/src/test/resources/models/mnist_scripted.mar      frontend/archive/src/test/resources/models/mnist_scripted.mar
R100    frontend/modelarchive/src/test/resources/models/mnist_traced.mar        frontend/archive/src/test/resources/models/mnist_traced.mar
R100    frontend/modelarchive/src/test/resources/models/noop-no-manifest.mar    frontend/archive/src/test/resources/models/noop-no-manifest.mar
R100    frontend/modelarchive/src/test/resources/models/noop-v1.0-config-tests.mar      frontend/archive/src/test/resources/models/noop-v1.0-config-tests.mar
R100    frontend/modelarchive/src/test/resources/models/noop.mar        frontend/archive/src/test/resources/models/noop.mar
R100    frontend/modelarchive/src/test/resources/models/noop_entrypntfunc.mar   frontend/archive/src/test/resources/models/noop_entrypntfunc.mar
R100    frontend/modelarchive/src/test/resources/models/noop_handlername.mar    frontend/archive/src/test/resources/models/noop_handlername.mar
R100    frontend/modelarchive/src/test/resources/models/noop_no_archive/noop/MAR-INF/MANIFEST.json      frontend/archive/src/test/resources/models/noop_no_archive/noop/MAR-INF/MANIFEST.json
A       frontend/archive/src/test/resources/models/noop_no_archive/noop/model.pt
R100    frontend/modelarchive/src/test/resources/models/noop_no_archive/noop/service.py frontend/archive/src/test/resources/models/noop_no_archive/noop/service.py
R100    frontend/modelarchive/src/test/resources/models/noop_v2.mar     frontend/archive/src/test/resources/models/noop_v2.mar
R100    frontend/modelarchive/src/test/resources/models/prediction-memory-error.mar     frontend/archive/src/test/resources/models/prediction-memory-error.mar
R100    frontend/modelarchive/src/test/resources/models/respheader-test.mar     frontend/archive/src/test/resources/models/respheader-test.mar
A       frontend/archive/src/test/resources/workflows/custom_invalid_python_dep.war
A       frontend/archive/src/test/resources/workflows/custom_python_dep.war
A       frontend/archive/src/test/resources/workflows/loading-memory-error.war
A       frontend/archive/src/test/resources/workflows/prediction-memory-error.war
A       frontend/archive/src/test/resources/workflows/smtest.war
R062    frontend/modelarchive/testng.xml        frontend/archive/testng.xml
M       frontend/gradle.properties
D       frontend/modelarchive/src/test/java/org/pytorch/serve/archive/CoverageTest.java



**issue_682_workflow_server_ensemble**

A       frontend/server/src/main/java/org/pytorch/serve/ensemble/Dag.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/DagExecutor.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/InvalidDAGException.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/Node.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/NodeOutput.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkFlow.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkflowInt.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkflowManifest.java
A       frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkflowModel.java



**issue_682_workflow_server_http**

M       frontend/server/src/main/java/org/pytorch/serve/grpcimpl/InferenceImpl.java
M       frontend/server/src/main/java/org/pytorch/serve/grpcimpl/ManagementImpl.java
M       frontend/server/src/main/java/org/pytorch/serve/http/HttpRequestHandler.java
M       frontend/server/src/main/java/org/pytorch/serve/http/HttpRequestHandlerChain.java
M       frontend/server/src/main/java/org/pytorch/serve/http/InvalidRequestHandler.java
R077    frontend/server/src/main/java/org/pytorch/serve/http/ApiDescriptionRequestHandler.java  frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ApiDescriptionRequestHandler.java
R092    frontend/server/src/main/java/org/pytorch/serve/http/InferenceRequestHandler.java       frontend/server/src/main/java/org/pytorch/serve/http/api/rest/InferenceRequestHandler.java
R090    frontend/server/src/main/java/org/pytorch/serve/http/ManagementRequestHandler.java      frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java
R087    frontend/server/src/main/java/org/pytorch/serve/http/PrometheusMetricsRequestHandler.java       frontend/server/src/main/java/org/pytorch/serve/http/api/rest/PrometheusMetricsRequestHandler.java
R099    frontend/server/src/main/java/org/pytorch/serve/http/DescribeModelResponse.java frontend/server/src/main/java/org/pytorch/serve/http/messages/DescribeModelResponse.java
R089    frontend/server/src/main/java/org/pytorch/serve/http/KFV1ModelReadyResponse.java        frontend/server/src/main/java/org/pytorch/serve/http/messages/KFV1ModelReadyResponse.java
R097    frontend/server/src/main/java/org/pytorch/serve/http/ListModelsResponse.java    frontend/server/src/main/java/org/pytorch/serve/http/messages/ListModelsResponse.java



**issue_682_workflow_server_wlm_utils**

M       frontend/server/build.gradle
M       frontend/server/src/main/java/org/pytorch/serve/ModelServer.java
M       frontend/server/src/main/java/org/pytorch/serve/ServerInitializer.java
M       frontend/server/src/main/java/org/pytorch/serve/job/RestJob.java
M       frontend/server/src/main/java/org/pytorch/serve/openapi/OpenApiUtils.java
M       frontend/server/src/main/java/org/pytorch/serve/snapshot/SnapshotManager.java
M       frontend/server/src/main/java/org/pytorch/serve/util/ApiUtils.java
M       frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java
M       frontend/server/src/main/java/org/pytorch/serve/util/JsonUtils.java
M       frontend/server/src/main/java/org/pytorch/serve/util/messages/EnvironmentUtils.java
M       frontend/server/src/main/java/org/pytorch/serve/wlm/Model.java
M       frontend/server/src/main/java/org/pytorch/serve/wlm/ModelManager.java
M       frontend/server/src/main/java/org/pytorch/serve/wlm/ModelVersionedRefs.java
M       frontend/server/src/main/java/org/pytorch/serve/wlm/WorkLoadManager.java



**issue_682_workflow_server_workflowcode**

A       frontend/server/src/main/java/org/pytorch/serve/workflow/WorkflowManager.java
A       frontend/server/src/main/java/org/pytorch/serve/workflow/api/http/WorkflowInferenceRequestHandler.java
A       frontend/server/src/main/java/org/pytorch/serve/workflow/api/http/WorkflowMgmtRequestHandler.java
A       frontend/server/src/main/java/org/pytorch/serve/workflow/messages/DescribeWorkflowResponse.java
A       frontend/server/src/main/java/org/pytorch/serve/workflow/messages/ListWorkflowResponse.java
A       frontend/server/src/main/java/org/pytorch/serve/workflow/messages/ModelRegistrationResult.java
A       frontend/server/src/main/java/org/pytorch/serve/workflow/messages/RegisterWorkflowRequest.java



**issue_682_workflow_server_test**


A       frontend/server/src/test/java/org/pytorch/serve/EnsembleTest.java
M       frontend/server/src/test/java/org/pytorch/serve/ModelServerTest.java
M       frontend/server/src/test/java/org/pytorch/serve/SnapshotTest.java
M       frontend/server/src/test/java/org/pytorch/serve/TestUtils.java
A       frontend/server/src/test/java/org/pytorch/serve/WorkflowTest.java
A       frontend/server/src/test/resources/classification_multiout_agg.py
A       frontend/server/src/test/resources/classification_multiout_agg.yaml
A       frontend/server/src/test/resources/classification_multiout_handler.py
A       frontend/server/src/test/resources/classification_multiout_workflow.yaml
M       frontend/server/src/test/resources/config.properties
M       frontend/server/src/test/resources/config_test_env.properties
A       frontend/server/src/test/resources/densenet_workflow.yaml
A       frontend/server/src/test/resources/densenet_workflow_handler.py
M       frontend/server/src/test/resources/inference_open_api.json
M       frontend/server/src/test/resources/management_open_api.json
M       frontend/server/src/test/resources/metrics_open_api.json
M       frontend/server/src/test/resources/snapshots/snapshot1.cfg
M       frontend/server/src/test/resources/snapshots/snapshot2.cfg
M       frontend/server/src/test/resources/snapshots/snapshot3.cfg
M       frontend/server/src/test/resources/snapshots/snapshot4.cfg
M       frontend/server/src/test/resources/snapshots/snapshot5.cfg
M       frontend/server/src/test/resources/snapshots/snapshot6.cfg
M       frontend/server/src/test/resources/snapshots/snapshot7.cfg
M       frontend/server/src/test/resources/snapshots/snapshot8.cfg
M       frontend/server/src/test/resources/snapshots/snapshot9.cfg
M       frontend/server/testng.xml
M       frontend/settings.gradle
M       frontend/tools/conf/pmd.xml



**issue_682_workflow_backend**

M       ts/arg_parser.py
M       ts/model_loader.py
M       ts/model_server.py
M       ts/protocol/otf_message_handler.py
A       ts/torch_handler/densenet_handler.py



**issue_682_workflow_scripts**

M       torchserve_sanity.py
M       ts_scripts/install_from_src.py
M       ts_scripts/install_utils
M       ts_scripts/sanity_utils.py
M       ts_scripts/tsutils.py
A       ts_scripts/workflow_archiver_utils.py



**issue_682_workflow_archiver_py**

A       workflow-archiver/.coveragerc
A       workflow-archiver/LICENSE
A       workflow-archiver/MANIFEST.in
A       workflow-archiver/PyPiDescription.rst
A       workflow-archiver/README.md
A       workflow-archiver/setup.py
A       workflow-archiver/workflow_archiver/__init__.py
A       workflow-archiver/workflow_archiver/arg_parser.py
A       workflow-archiver/workflow_archiver/manifest_components/__init__.py
A       workflow-archiver/workflow_archiver/manifest_components/manifest.py
A       workflow-archiver/workflow_archiver/manifest_components/workflow.py
A       workflow-archiver/workflow_archiver/tests/integ_tests/__init__.py
A       workflow-archiver/workflow_archiver/tests/integ_tests/configuration.json
A       workflow-archiver/workflow_archiver/tests/integ_tests/resources/test_handler.py
A       workflow-archiver/workflow_archiver/tests/integ_tests/resources/test_spec.yaml
A       workflow-archiver/workflow_archiver/tests/integ_tests/test_integration_workflow_archiver.py
A       workflow-archiver/workflow_archiver/tests/pylintrc
A       workflow-archiver/workflow_archiver/tests/unit_tests/test_version.py
A       workflow-archiver/workflow_archiver/tests/unit_tests/test_workflow_packaging.py
A       workflow-archiver/workflow_archiver/tests/unit_tests/test_workflow_packaging_utils.py
A       workflow-archiver/workflow_archiver/version.py
A       workflow-archiver/workflow_archiver/version.txt
A       workflow-archiver/workflow_archiver/workflow_archiver_error.py
A       workflow-archiver/workflow_archiver/workflow_packaging.py
A       workflow-archiver/workflow_archiver/workflow_packaging_utils.py

**Comment by dhanainme** [comment]
## Steps to perform a simple test in a DL AMI


** Clone Repo / Install Torchserve / WF Archiver / Model Archiver**

```
mkdir workflow
cd workflow/
git clone https://github.com/pytorch/serve
cd serve
git checkout issue_682
pip install . 


cd model-archiver/
pip install . 
cd ..

cd workflow-archiver/
pip install 
cd ..

```

** Create WF / Model Archives and move to a Model store - /tmp/workflow_test/**

```
cd examples/Workflows/densenet_image_classifier_pipeline/densenet_model

wget https://download.pytorch.org/models/densenet161-8d451a50.pth

torch-model-archiver --model-name densenet_wf --version 1.0 --model-file model.py --serialized-file densenet161-8d451a50.pth --handler densenet_handler.py

cd .. 

torch-workflow-archiver --workflow-name densenet --spec-file densenet_workflow.yaml --handler densenet_workflow_handler.py --extra-files index_to_name.json

mkdir /tmp/workflow_test/
mv densenet.war /tmp/workflow_test/
mv densenet_model/densenet_wf.mar /tmp/workflow_test/

```


** Start TS; Test WF Register; List; Inference; Model Register **
```
torchserve --start --model-store /tmp/workflow_test/ --workflow-store /tmp/workflow_test/

curl -X POST "http://127.0.0.1:8081/workflows?url=densenet.war"

curl http://localhost:8081/workflows/

curl http://127.0.0.1:8081/models/

curl http://localhost:8081/workflows/densenet

curl -X POST "http://127.0.0.1:8081/workflows?url=densenet.war"

curl http://127.0.0.1:8080/wfpredict/densenet -T kitten.jpg

```

Issue 
https://github.com/pytorch/serve/issues/682

PR 
https://github.com/pytorch/serve/pull/785


Workflow Spec Yaml
* serve/examples/Workflows/densenet_image_classifier_pipeline/densenet_workflow.yaml


Docs : 
* https://github.com/pytorch/serve/blob/issue_682/docs/workflows.md
* https://github.com/pytorch/serve/blob/issue_682/docs/workflow_management_api.md
* https://github.com/pytorch/serve/blob/issue_682/docs/workflow_inference_api.md



Known issues : 
https://github.com/pytorch/serve/blob/issue_682/docs/workflows.md#known-issues


WF Scenarios planned for support : 
https://docs.google.com/spreadsheets/d/1x_Rj5xczANznVRJBaMrU0Wkhy-z-uWasrS4bzkQ7s3c/edit?ts=5f882aa5#gid=0
Only Pipeline scenario supported now.

File: frontend/server/src/main/java/org/pytorch/serve/ensemble/Node.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package org.pytorch.serve.ensemble;
+
+public class Node {
+
+    private String name;
+    private String parentName;
+    private WorkflowModel workflowModel;
```
Comment (dhanainme) [inline frontend/server/src/main/java/org/pytorch/serve/ensemble/Node.java:7]: final / Lombok

File: frontend/server/src/main/java/org/pytorch/serve/ensemble/Dag.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+        ArrayList<String> topoSortedList = new ArrayList<>();
+        DagExecutor de = new DagExecutor(this);
+        de.execute(null, topoSortedList);
+        if (topoSortedList.size() != nodes.size()) {
+            throw new InvalidDAGException("Not a valid DAG");
+        }
+        return topoSortedList;
+    }
+}
```
Comment (dhanainme) [inline frontend/server/src/main/java/org/pytorch/serve/ensemble/Dag.java:112]: Refactor to use Guava Graph lib - https://guava.dev/releases/23.0/api/docs/com/google/common/graph/Graph.html

File: frontend/server/src/main/java/org/pytorch/serve/ensemble/DagExecutor.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            logger.error("Model not found.");
+            logger.error(e.getMessage());
+            throw e;
+        } catch (ModelVersionNotFoundException e) {
+            logger.error("Model version not found.");
+            logger.error(e.getMessage());
+            throw e;
+        }
+    }
+}
```
Comment (dhanainme) [inline frontend/server/src/main/java/org/pytorch/serve/ensemble/DagExecutor.java:161]: ListenableFuture may be a more appropriate 

https://guava.dev/releases/21.0/api/docs/com/google/common/util/concurrent/ListenableFuture.html

https://stackoverflow.com/questions/19138212/how-to-implement-a-dag-like-scheduler-in-java has a good example.

File: frontend/server/src/main/java/org/pytorch/serve/ensemble/NodeOutput.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        this.nodeName = nodeName;
+    }
+
+    public Object getData() {
+        return data;
+    }
+
+    public void setData(Object data) {
+        this.data = data;
+    }
```
Comment (dhanainme) [inline frontend/server/src/main/java/org/pytorch/serve/ensemble/NodeOutput.java:26]: Lombok

File: frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkflowManifest.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+import com.google.gson.annotations.SerializedName;
+
+public class WorkflowManifest {
+
+    private String createdOn;
+    private String description;
+    private String archiverVersion;
+    private RuntimeType runtime;
+    private Workflow workflow;
```
Comment (dhanainme) [inline frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkflowManifest.java:11]: final / lombok

File: frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkflowModel.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+public class WorkflowModel {
+
+    private String name;
+    private String url;
+    private int minWorkers;
+    private int maxWorkers;
+    private int batchSize;
+    private int maxBatchDelay;
+    private String handler;
```
Comment (dhanainme) [inline frontend/server/src/main/java/org/pytorch/serve/ensemble/WorkflowModel.java:11]: final / lombok

**Review by dhanainme — CHANGES_REQUESTED** [review]
Reviewed Ensemble DAG related changes

* Lombok / final (To be fixed)
* Usage of Guava / ListenableFuture / Yaml lib (Create issues)

File: workflow-archiver/workflow_archiver/workflow_packaging_utils.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        :return:
+        """
+        unwanted_dirs = {'__MACOSX', '__pycache__'}
+
+        for root, directories, files in os.walk(path):
+            # Filter directories
+            directories[:] = [d for d in directories if WorkflowExportUtils.directory_filter(d, unwanted_dirs)]
+            for f in files:
+                file_path = os.path.join(root, f)
+                dst.write(file_path, os.path.relpath(file_path, path))
```
Comment (dhanainme) [inline workflow-archiver/workflow_archiver/workflow_packaging_utils.py:146]: May have to explicitly document this behavior on what files get archived.

File: workflow-archiver/workflow_archiver/workflow_packaging_utils.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        if not re.match(r'^[A-Za-z0-9][A-Za-z0-9_\-.]*$', workflow_name):
+            raise WorkflowArchiverError("Workflow name contains special characters.\n"
+                                        "The allowed regular expression filter for workflow "
+                                        "name is: ^[A-Za-z0-9][A-Za-z0-9_\\-.]*$")
+
+    @staticmethod
+    def validate_inputs(workflow_name, export_path):
+        WorkflowExportUtils.check_workflow_name_regex_or_exit(workflow_name)
+        if not os.path.isdir(os.path.abspath(export_path)):
+            raise WorkflowArchiverError("Given export-path {} is not a directory. "
```
Comment (dhanainme) [inline workflow-archiver/workflow_archiver/workflow_packaging_utils.py:197]: Nit : Separate methods for single method calls which can be grouped.

**Review by dhanainme — CHANGES_REQUESTED** [review]
Reviewed code for archiver CLI 

Callouts 

* Document behavior for what gets zipped up explicitly. 

Integ test code has not been reviewed yet.

File: ts/arg_parser.py [hunk 9]
```diff
@@ -33,6 +33,10 @@ def ts_parser():
                             required=False,
                             dest='model_store',
                             help='Model store location from where local or default models can be loaded')
+        parser.add_argument('--workflow-store',
+                            required=False,
+                            dest='workflow_store',
+                            help='Workflow store location from where local or default workflows can be loaded')
```
Comment (dhanainme) [inline ts/arg_parser.py:39]: Q : Why 2 stores ? Why not just 1 dir for both models & workflow.

**Review by dhanainme — CHANGES_REQUESTED** [review]

File: examples/Workflows/densenet_image_classifier_pipeline/densenet_model/densenet_handler.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
+"""
```
Comment (HamidShojanazeri) [inline examples/Workflows/densenet_image_classifier_pipeline/densenet_model/densenet_handler.py:1]: @dhanainme We need to document/add user guide for the differences of the handler for models(nodes) in the workflow with regular handlers we have in the non-workflow use-cases.

File: docs/workflows.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+TorchServe can be used for serving ensemble of models & functions (python) through Workflow APIs. 
+
+It utilizes [REST based APIs](rest_api.md) for workflow management and predictions.
+
+A Workflow is served on TorchServe using a workflow-archive(.war) which comprises of following: 
+
+## Workflow Specification file
+
+A workflow specification is a YAML file which provides the details of the models to be executed and a DAG for defining data flow.
```
Comment (HamidShojanazeri) [inline docs/workflows.md:11]: @dhanainme Please add some clarification on how the gpu assignment would be handled here.

File: ts/torch_handler/densenet_handler.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
+"""
```
Comment (HamidShojanazeri) [inline ts/torch_handler/densenet_handler.py:1]: @dhanainme is this planned to be aded as default handler? or having it in the example would suffice.

**Review by HamidShojanazeri — COMMENTED** [review]
@dhanainme Please refer to inline comments.

**Comment by HamidShojanazeri** [comment]
@dhanainme I was wondering in case having multiple models in the DAG, then we will have one worker-manager that have a copy of all the models + CUDA runtime?

**Comment by msaroufim** [comment]
Hi I'm new so lacking context

This PR seems to be about adding sequential workflows but just trying to understand how this differs in scope from a regular handler? https://github.com/pytorch/serve/blob/master/examples/Huggingface_Transformers/Transformer_handler_generalized.py

Is the intention that this PR makes it easier to support non sequential workflows in the future?

**Comment by HamidShojanazeri** [comment]
> Hi I'm new so lacking context
> 
> This PR seems to be about adding sequential workflows but just trying to understand how this differs in scope from a regular handler? https://github.com/pytorch/serve/blob/master/examples/Huggingface_Transformers/Transformer_handler_generalized.py
> 
> Is the intention that this PR makes it easier to support non sequential workflows in the future?

@msaroufim Initially, we were targeting with sequential pipeline support for this release and adding support for other workflow DAGs in next releases. The current status of the PR, merged into release_0.4.0 branch supports non-sequential as well, although we are looking into improvements in future releases.

**Pull request merged**